nav2_costmap_2d/include/nav2_costmap_2d/costmap_filters/costmap_filter.hpp

```diff
@@ -0,0 +1,135 @@
+ *     with the distribution.
+ *   * Neither the name of the <ORGANIZATION> nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+#ifndef NAV2_COSTMAP_2D__COSTMAP_FILTER_HPP_
+#define NAV2_COSTMAP_2D__COSTMAP_FILTER_HPP_
+
+#include "rclcpp/rclcpp.hpp"
+
+#include <string>
+
+#include "nav2_costmap_2d/costmap_layer.hpp"
+
+namespace nav2_costmap_2d
+{
+
+class CostmapFilter : public CostmapLayer
+{
+public:
+  CostmapFilter();
+  ~CostmapFilter();
+
+  /** Layer API **/
+  virtual void onInitialize();
+  virtual void updateBounds(
+    double robot_x, double robot_y, double robot_yaw,
+    double * min_x, double * min_y, double * max_x, double * max_y);
+  virtual void updateCosts(
+    nav2_costmap_2d::Costmap2D & master_grid,
+    int min_i, int min_j, int max_i, int max_j);
+  virtual void matchSize();
+
+  virtual void activate();
+  virtual void deactivate();
+  virtual void reset();
+
+  /** CostmapFilter API **/
+  /**
+   * @brief: Loads map filter. Creates subscriptions to a map filter related topics
+   * @param: Name of semantic info topic for map filter
+   */
+  virtual void loadFilter(
+    const std::string semantic_info_topic) = 0;
+
+  /**
+   * @brief: An algorithm for how to use that map's information. Fills the Costmap2D with
+   *         calculated data and makes an action based on processed data
+   * @param: Reference to a master costmap2d
+   * @param: Low window map boundary OX
+   * @param: Low window map boundary OY
+   * @param: High window map boundary OX
+   * @param: High window map boundary OY
+   * @param: Robot position x
+   * @param: Robot position y
+   * @param: Robot orientation
+   */
+  virtual void process(
+    nav2_costmap_2d::Costmap2D & master_grid,
+    int min_i, int min_j, int max_i, int max_j,
+    double robot_x, double robot_y, double robot_yaw) = 0;
+
+  /**
+   * @brief: Unloads map filter. Stops all subscriptions
+   */
+  virtual void unloadFilter() = 0;
```

Comment:
I don't think unload is the right word for what you mean, I think its reset.

Reply:
Okay, will rename it to `resetFilter()`.

Reply:
Why not just `reset()`?

Reply:
I do not want a confusion between Layer API and CostmapFilter API while using same names of base methods.

Reply:
Fair enough - please proceed with that.

Reply:
Done, https://github.com/ros-planning/navigation2/commit/fcf337566456677d0a6892b8539e902dfd086994#diff-cffa086d2532b26fb31fc174a5b5878cR74 and https://github.com/ros-planning/navigation2/commit/fcf337566456677d0a6892b8539e902dfd086994#diff-cffa086d2532b26fb31fc174a5b5878cR97